### PR TITLE
Feature/atr 139 dev forbid static graph members

### PR DESCRIPTION
**Pull request by SENya1990**
Overview:
- implemented check that graph does not contain static:
           - views
           - actions
           - mutable fields
           - mutable properties
- added two code fixes:
           - remove static modifier from the static graph member
           - only for static mutable fields and properties - make them read only
- added new helpers to determine if symbol or syntax node is read only
- added helper to work with diagnostic properties passed to code fixes
- changed logging of the unobserved task exceptions in Visual Studio. Filter out timeout exceptions to reduce contention for VS UI thread. 
- added unit tests
- added example to the demo solution

### CODE REVIEW

File: src/Acuminator/Acuminator.Analyzers/Resources.resx [hunk 1]
```diff
@@ -643,4 +643,37 @@
   <data name="PX1072Title_CreateSharedGraphVariable" xml:space="preserve">
     <value>Creating a graph for a BQL query is a resource consuming operation. To decrease the number of these operations, create a variable with a shared graph that can be reused by BQL queries.</value>
   </data>
+  <data name="PX1062MessageFormat" xml:space="preserve">
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/Resources.resx:646]: @EAndrosova  please review the diagnostic message template: `Graphs and graph extensions cannot declare static {0}`
The list of possible format arguments (they are also here in resources):
- mutable fields
- mutable properties
- views
- actions

File: src/Acuminator/Acuminator.Analyzers/Resources.resx [hunk 2]
```diff
@@ -643,4 +643,37 @@
...
+    <value>property</value>
+  </data>
+  <data name="PX1062MessageFormatArg_Views" xml:space="preserve">
+    <value>views</value>
+  </data>
+  <data name="PX1062MessageFormatArg_Actions" xml:space="preserve">
+    <value>actions</value>
+  </data>
+  <data name="PX1062FixMakeNonStaticFormat" xml:space="preserve">
+    <value>Remove static from the {0}</value>
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/Resources.resx:671]: @EAndrosova  please review the code fix message: `Remove static from the {0}`
The possible format arguments (stored here in the resources):
- field
- property
- view
- action

File: src/Acuminator/Acuminator.Analyzers/Resources.resx [hunk 3]
```diff
@@ -643,4 +643,37 @@
...
+    <value>Graphs and graph extensions cannot declare static {0}.</value>
+  </data>
+  <data name="PX1062MessageFormatArg_Fields" xml:space="preserve">
+    <value>mutable fields</value>
+  </data>
+  <data name="PX1062MessageFormatArg_Properties" xml:space="preserve">
+    <value>mutable properties</value>
+  </data>
+  <data name="PX1062FixMakeReadOnlyFormat" xml:space="preserve">
+    <value>Make the {0} read-only</value>
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/Resources.resx:656]: @EAndrosova  please review the code fix message: `Make the {0} read-only`
The possible format arguments (stored here in the resources):
- field
- property
- view
- action

**Review by dropsonic — APPROVED** [review]

File: src/Acuminator/Acuminator.Analyzers/DiagnosticsShortName.resx [hunk 4]
```diff
@@ -372,4 +372,7 @@
   <data name="PX1056IsActiveForGraphMethod" xml:space="preserve">
     <value>PXGraphCreationInIsActiveForGraphMethod</value>
   </data>
+  <data name="PX1062" xml:space="preserve">
+    <value>StaticPropertyOrFieldInGraph</value>
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/DiagnosticsShortName.resx:376]: @EAndrosova  please check the diagnostic short name

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 1 above)
Comment (EAndrosova) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:646]: @SENya1990  I think we need an object in this sentence. Can we change it to "You cannot declare static {0} in graphs or graph exntesions."

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 2 above)
Comment (EAndrosova) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:671]: @SENya1990  I think this sentence should be a bit more specific: "Remove the static modifier from the {0}"
Also, can this message specify the exact name of the field? For example, _Remove the static modifier for the InvoiceNbr field._

File: src/Acuminator/Acuminator.Analyzers/DiagnosticsShortName.resx (hunk 4 above)
Comment (EAndrosova) [reply src/Acuminator/Acuminator.Analyzers/DiagnosticsShortName.resx:376]: approved

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 3 above)
Comment (EAndrosova) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:656]: To follow some grammar rules, please change to "Add the readonly modifier to the field" or "Make the field read-only."

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 1 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:646]: Yes, I'll change

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 3 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:656]: The current code fix message is exactly like suggested except the dot in the end. 
I think we should not place dots at the end of diagnostic and code fix messages, I explained the reasons in this comment:
https://github.com/Acumatica/Acuminator/pull/462#issuecomment-1306173979

Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:656]: After discussion it was decided to follow MS style for  diagnostic messages:

- No dots for single sentences
- Dots at the end of all sentences if there are two or more sentences

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 2 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:671]: After discussion it was decided to follow MS style for diagnostic messages:

-     No dots for single sentences
-     Dots at the end of all sentences if there are two or more sentences

Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx]: done

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 1 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx:646]: done

**Review by EAndrosova — APPROVED** [review]

**Pull request merged**